packages/flutter/lib/src/material/expansion_panel.dart

```diff
@@ -106,6 +117,243 @@ class ExpansionPanel {
+  /// class CustomAnimatedIcon extends StatefulWidget{
+  ///   const CustomAnimatedIcon(
+  ///     this.isExpanded,
+  ///     this.duration,
+  ///   );
+  ///
+  ///   final bool isExpanded;
+  ///   final Duration duration;
+  ///
+  ///   @override
+  ///   _CustomAnimatedIconState createState() => _CustomAnimatedIconState();
+  /// }
+  ///
+  /// class _CustomAnimatedIconState extends State<CustomAnimatedIcon> with SingleTickerProviderStateMixin {
+  ///   late AnimationController animationController;
+  ///
+  ///   @override
+  ///   void initState() {
+  ///     super.initState();
+  ///     animationController = AnimationController(
+  ///       vsync: this,
+  ///       duration: widget.duration,
+  ///     );
+  ///   }
+  ///
+  ///   @override
+  ///   void didUpdateWidget(CustomAnimatedIcon oldWidget) {
+  ///     if (widget.isExpanded != oldWidget.isExpanded) {
+  ///       if (widget.isExpanded) {
+  ///         animationController.forward();
+  ///       } else {
+  ///         animationController.reverse();
+  ///       }
+  ///     }
+  ///     super.didUpdateWidget(oldWidget);
+  ///   }
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return AnimatedIcon(
+  ///       icon: AnimatedIcons.menu_close,
+  ///       progress: animationController,
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// ```dart
+  /// bool _isExpanded = false;
+  ///
+  /// @override
+  /// Widget build(BuildContext context) {
+  ///   return SingleChildScrollView(
+  ///     child: ExpansionPanelList(
+  ///       expansionCallback: (int index, bool isExpanded) {
+  ///         setState(() {
+  ///           _isExpanded = !isExpanded;
+  ///         });
+  ///       },
+  ///       children: <ExpansionPanel>[
+  ///         ExpansionPanel(
+  ///           isExpanded: _isExpanded,
+  ///           headerBuilder: (BuildContext context, bool isExpanded) {
+  ///             return const ListTile(title: Text('Header Text'));
+  ///           },
+  ///           expandIconBuilder: (
+  ///             BuildContext context,
+  ///             bool isExpanded,
+  ///             VoidCallback handlePressed,
+  ///             Duration animationDuration,
+  ///           ) {
+  ///             return InkWell(
+  ///               customBorder: const CircleBorder(),
+  ///               onTap: () {
+  ///                 handlePressed();
+  ///               },
+  ///               child: Padding(
+  ///                 padding: const EdgeInsets.all(12.0),
+  ///                 child: CustomAnimatedIcon(isExpanded, animationDuration),
+  ///               ),
+  ///             );
+  ///           },
+  ///           body: const ListTile(
+  ///             title: Text('Title Text'),
+  ///             subtitle: Text('Subtitle Text'),
+  ///           ),
+  ///         ),
+  ///       ],
+  ///     ),
+  ///   );
+  /// }
+  /// ```
+  /// {@end-tool}
+  final ExpansionPanelIconBuilder? expandIconBuilder;
```

Comment:
This seems to generic, based on the original issue its seems developers only want to be able to customize the icon. While this API is powerful, there is some boiler plate code they need to write.

Maybe we should only expose the expandedIcon and collapsedIcon in the constructor.  WDYT?

Reply:
I think that maybe we can add both options to devs. Most of the use cases will be from people that just wanna change the icon, but there will be some thinking about a custom widget (like the checkbox on the sample) and for them, having this as default instead of the need to create a full custom panel from the start, would be very helpful. @chunhtai, we could have another parameter  `expansionPanelIcon` to achieve the first option, and this for complex cases (defining both should throw an assert).